controllers/notificationpolicy_controller.go

```diff
@@ -222,6 +336,132 @@ func (r *GrafanaNotificationPolicyReconciler) SetupWithManager(mgr ctrl.Manager)
 			}
 			return requests
 		})).
+		Watches(&grafanav1beta1.GrafanaNotificationPolicyRoute{}, handler.EnqueueRequestsFromMapFunc(func(ctx context.Context, o client.Object) []reconcile.Request {
+			// resync all notification policies that have a routeSelector that matches the routes labels
+			nps := &grafanav1beta1.GrafanaNotificationPolicyList{}
+			if err := r.List(ctx, nps); err != nil {
+				r.Log.Error(err, "failed to fetch notification policies for watch mapping")
+				return nil
+			}
+			requests := []reconcile.Request{}
+			for _, np := range nps.Items {
+				if !hasRouteSelector(np.Spec.Route) {
+					continue
+				}
+
+				if np.GetNamespace() != o.GetNamespace() && !np.IsCrossNamespaceImportAllowed() {
+					continue
+				}
+
+				allRouteSelectors := getRouteSelectors(np.Spec.Route)
+
+				for _, routeSelector := range allRouteSelectors {
+					selector, err := metav1.LabelSelectorAsSelector(routeSelector)
+					if err != nil {
+						r.Log.Error(err, "failed to create selector from RouteSelector")
+						continue
+					}
+
+					if selector.Matches(labels.Set(o.GetLabels())) {
+						requests = append(requests,
+							reconcile.Request{
+								NamespacedName: types.NamespacedName{
+									Name:      np.Name,
+									Namespace: np.Namespace,
+								},
+							})
+					}
+				}
+			}
+			return requests
+		})).
 		WithEventFilter(ignoreStatusUpdates()).
 		Complete(r)
 }
+
+// getMatchingNotificationPolicyRoutes retrieves all GrafanaNotificationPolicyRoutes for the given labelSelector
+// results will be limited to namespace when specified
+func getMatchingNotificationPolicyRoutes(ctx context.Context, k8sClient client.Client, labelSelector *metav1.LabelSelector, namespace *string) (*v1beta1.GrafanaNotificationPolicyRouteList, error) {
```

Comment:
Same suggestion regarding call signature. This can be a struct function for the reconciler